In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv(r"C:\Users\Admin\OneDrive\Desktop\climate-intelligence-prototype\data\climate_data.csv")

df.head()

,Date,City,State,Temperature_Max (°C),Temperature_Min (°C),Temperature_Avg (°C),Humidity (%),Rainfall (mm),Wind_Speed (km/h),AQI,AQI_Category,Pressure (hPa),Cloud_Cover (%)
0,2024-01-01,Mumbai,Maharashtra,32.5,18.0,25.2,77.6,0.0,3.3,259,Poor,1020.3,62.1
1,2024-01-01,Delhi,Delhi,25.4,10.7,18.1,84.1,0.0,9.0,130,Moderate,1008.4,46.0
2,2024-01-01,Bengaluru,Karnataka,37.2,30.8,34.0,49.0,3.7,6.6,54,Satisfactory,1008.0,61.3
3,2024-01-01,Chennai,Tamil Nadu,37.2,30.4,33.8,34.2,9.5,9.0,176,Moderate,993.4,70.0
4,2024-01-01,Kolkata,West Bengal,27.4,17.5,22.5,32.2,9.1,9.2,97,Satisfactory,1008.2,56.9


In [3]:
print("Dataset Shape:", df.shape)

print("\nDataset Information:")
df.info()

print("\nMissing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:", df.duplicated().sum())

Dataset Shape: (7310, 13)

Dataset Information:
<class 'pandas.DataFrame'>
RangeIndex: 7310 entries, 0 to 7309
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Date                  7310 non-null   str    
 1   City                  7310 non-null   str    
 2   State                 7310 non-null   str    
 3   Temperature_Max (°C)  7310 non-null   float64
 4   Temperature_Min (°C)  7310 non-null   float64
 5   Temperature_Avg (°C)  7310 non-null   float64
 6   Humidity (%)          7310 non-null   float64
 7   Rainfall (mm)         7310 non-null   float64
 8   Wind_Speed (km/h)     7310 non-null   float64
 9   AQI                   7310 non-null   int64  
 10  AQI_Category          7310 non-null   str    
 11  Pressure (hPa)        7310 non-null   float64
 12  Cloud_Cover (%)       7310 non-null   float64
dtypes: float64(8), int64(1), str(4)
memory usage: 742.6 KB

Missing Values:
Date          

In [4]:
# Check missing values
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

print("\nTotal missing values:", missing_values.sum())

# Check duplicate rows
duplicate_count = df.duplicated().sum()

print("\nTotal duplicate rows:", duplicate_count)

Missing values in each column:
Date                    0
City                    0
State                   0
Temperature_Max (°C)    0
Temperature_Min (°C)    0
Temperature_Avg (°C)    0
Humidity (%)            0
Rainfall (mm)           0
Wind_Speed (km/h)       0
AQI                     0
AQI_Category            0
Pressure (hPa)          0
Cloud_Cover (%)         0
dtype: int64

Total missing values: 0

Total duplicate rows: 0


In [5]:
# Convert Date column to datetime format
df["Date"] = pd.to_datetime(df["Date"], errors="coerce")

print("Date datatype:", df["Date"].dtype)

print("\nInvalid dates after conversion:", df["Date"].isnull().sum())

Date datatype: datetime64[us]

Invalid dates after conversion: 0


In [6]:
# Check for logically invalid values

print("Humidity values outside 0-100:")
print(((df["Humidity (%)"] < 0) | (df["Humidity (%)"] > 100)).sum())

print("\nCloud cover values outside 0-100:")
print(((df["Cloud_Cover (%)"] < 0) | (df["Cloud_Cover (%)"] > 100)).sum())

print("\nNegative rainfall values:")
print((df["Rainfall (mm)"] < 0).sum())

print("\nNegative wind speed values:")
print((df["Wind_Speed (km/h)"] < 0).sum())

print("\nInvalid temperature relationship (Max < Min):")
print((df["Temperature_Max (°C)"] < df["Temperature_Min (°C)"]).sum())

print("\nNegative AQI values:")
print((df["AQI"] < 0).sum())

print("\nInvalid pressure values (<= 0):")
print((df["Pressure (hPa)"] <= 0).sum())

Humidity values outside 0-100:
0

Cloud cover values outside 0-100:
0

Negative rainfall values:
0

Negative wind speed values:
0

Invalid temperature relationship (Max < Min):
0

Negative AQI values:
0

Invalid pressure values (<= 0):
0


In [7]:
# Check outliers using the IQR method

numeric_columns = [
    "Temperature_Max (°C)",
    "Temperature_Min (°C)",
    "Temperature_Avg (°C)",
    "Humidity (%)",
    "Rainfall (mm)",
    "Wind_Speed (km/h)",
    "AQI",
    "Pressure (hPa)",
    "Cloud_Cover (%)"
]

outlier_summary = {}

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = ((df[column] < lower_bound) | (df[column] > upper_bound)).sum()

    outlier_summary[column] = outliers

print("Outlier count for each numerical column:")
for column, count in outlier_summary.items():
    print(f"{column}: {count}")

Outlier count for each numerical column:
Temperature_Max (°C): 0
Temperature_Min (°C): 0
Temperature_Avg (°C): 0
Humidity (%): 0
Rainfall (mm): 1040
Wind_Speed (km/h): 0
AQI: 0
Pressure (hPa): 0
Cloud_Cover (%): 0


In [8]:
Q1 = df["Rainfall (mm)"].quantile(0.25)
Q3 = df["Rainfall (mm)"].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

print("\nRainfall statistics:")
print(df["Rainfall (mm)"].describe())

print("\nMaximum rainfall:", df["Rainfall (mm)"].max())


Q1: 0.0
Q3: 6.3
IQR: 6.3
Lower Bound: -9.45
Upper Bound: 15.75

Rainfall statistics:
count    7310.000000
mean        8.231300
std        17.758208
min         0.000000
25%         0.000000
50%         0.000000
75%         6.300000
max        79.900000
Name: Rainfall (mm), dtype: float64

Maximum rainfall: 79.9


In [9]:
# Create time-based features from Date

df["Year"] = df["Date"].dt.year
df["Month"] = df["Date"].dt.month
df["Day"] = df["Date"].dt.day
df["Day_of_Week"] = df["Date"].dt.dayofweek
df["Day_Name"] = df["Date"].dt.day_name()

df[["Date", "Year", "Month", "Day", "Day_of_Week", "Day_Name"]].head()

,Date,Year,Month,Day,Day_of_Week,Day_Name
0,2024-01-01,2024,1,1,0,Monday
1,2024-01-01,2024,1,1,0,Monday
2,2024-01-01,2024,1,1,0,Monday
3,2024-01-01,2024,1,1,0,Monday
4,2024-01-01,2024,1,1,0,Monday


In [10]:
# Create weather-based derived features

# Temperature difference between maximum and minimum
df["Temperature_Range (°C)"] = (
    df["Temperature_Max (°C)"] - df["Temperature_Min (°C)"]
)

# Average temperature difference from maximum temperature
df["Max_Avg_Temp_Difference (°C)"] = (
    df["Temperature_Max (°C)"] - df["Temperature_Avg (°C)"]
)

# Average temperature difference from minimum temperature
df["Avg_Min_Temp_Difference (°C)"] = (
    df["Temperature_Avg (°C)"] - df["Temperature_Min (°C)"]
)

# Combined humidity and temperature indicator
df["Humidity_Temperature_Index"] = (
    df["Humidity (%)"] * df["Temperature_Avg (°C)"]
)

df[
    [
        "Temperature_Max (°C)",
        "Temperature_Min (°C)",
        "Temperature_Avg (°C)",
        "Temperature_Range (°C)",
        "Max_Avg_Temp_Difference (°C)",
        "Avg_Min_Temp_Difference (°C)",
        "Humidity_Temperature_Index"
    ]
].head()

,Temperature_Max (°C),Temperature_Min (°C),Temperature_Avg (°C),Temperature_Range (°C),Max_Avg_Temp_Difference (°C),Avg_Min_Temp_Difference (°C),Humidity_Temperature_Index
0,32.5,18.0,25.2,14.5,7.3,7.2,1955.52
1,25.4,10.7,18.1,14.7,7.3,7.4,1522.21
2,37.2,30.8,34.0,6.4,3.2,3.2,1666.00
3,37.2,30.4,33.8,6.8,3.4,3.4,1155.96
4,27.4,17.5,22.5,9.9,4.9,5.0,724.50


In [11]:
# Check correlations among numerical features

correlation_matrix = df.select_dtypes(include=["number"]).corr()

correlation_matrix.round(2)


,Temperature_Max (°C),Temperature_Min (°C),Temperature_Avg (°C),Humidity (%),Rainfall (mm),Wind_Speed (km/h),AQI,Pressure (hPa),Cloud_Cover (%),Year,Month,Day,Day_of_Week,Temperature_Range (°C),Max_Avg_Temp_Difference (°C),Avg_Min_Temp_Difference (°C),Humidity_Temperature_Index
Temperature_Max (°C),1.00,0.90,0.97,0.01,-0.02,-0.02,0.01,-0.00,0.00,0.01,0.01,0.03,0.00,-0.01,-0.01,-0.01,0.54
Temperature_Min (°C),0.90,1.00,0.98,0.00,-0.02,-0.02,0.01,-0.01,-0.01,0.01,0.00,0.03,-0.01,-0.45,-0.45,-0.45,0.54
Temperature_Avg (°C),0.97,0.98,1.00,0.00,-0.02,-0.02,0.01,-0.00,-0.00,0.01,0.01,0.03,-0.00,-0.25,-0.25,-0.25,0.55
Humidity (%),0.01,0.00,0.00,1.00,-0.00,-0.01,-0.01,0.00,0.03,-0.01,-0.00,-0.01,-0.01,0.01,0.01,0.00,0.82
Rainfall (mm),-0.02,-0.02,-0.02,-0.00,1.00,-0.01,0.00,0.00,0.01,-0.01,-0.01,0.01,0.02,0.01,0.01,0.01,-0.01
Wind_Speed (km/h),-0.02,-0.02,-0.02,-0.01,-0.01,1.00,-0.01,0.01,-0.01,0.02,0.01,-0.00,-0.00,0.01,0.01,0.01,-0.02
AQI,0.01,0.01,0.01,-0.01,0.00,-0.01,1.00,-0.00,-0.01,0.01,0.01,0.01,-0.02,-0.01,-0.01,-0.01,0.00
Pressure (hPa),-0.00,-0.01,-0.00,0.00,0.00,0.01,-0.00,1.00,-0.00,-0.01,-0.00,0.01,-0.01,0.01,0.01,0.01,0.00
Cloud_Cover (%),0.00,-0.01,-0.00,0.03,0.01,-0.01,-0.01,-0.00,1.00,-0.00,-0.01,-0.00,0.01,0.02,0.02,0.02,0.02
Year,0.01,0.01,0.01,-0.01,-0.01,0.02,0.01,-0.01,-0.00,1.00,0.00,-0.00,0.00,0.01,0.01,0.01,-0.00


In [12]:
# Remove redundant temperature difference features

df.drop(
    columns=[
        "Max_Avg_Temp_Difference (°C)",
        "Avg_Min_Temp_Difference (°C)"
    ],
    inplace=True
)

print("Removed redundant features:")
print("Max_Avg_Temp_Difference (°C)")
print("Avg_Min_Temp_Difference (°C)")

print("\nRemaining columns:")
print(df.columns.tolist())

Removed redundant features:
Max_Avg_Temp_Difference (°C)
Avg_Min_Temp_Difference (°C)

Remaining columns:
['Date', 'City', 'State', 'Temperature_Max (°C)', 'Temperature_Min (°C)', 'Temperature_Avg (°C)', 'Humidity (%)', 'Rainfall (mm)', 'Wind_Speed (km/h)', 'AQI', 'AQI_Category', 'Pressure (hPa)', 'Cloud_Cover (%)', 'Year', 'Month', 'Day', 'Day_of_Week', 'Day_Name', 'Temperature_Range (°C)', 'Humidity_Temperature_Index']


In [13]:
# Final validation of the cleaned dataset

print("Final dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

print("\nFirst 5 rows:")
display(df.head())

print("\nFinal columns:")
print(df.columns.tolist())

Final dataset shape: (7310, 20)

Missing values:
0

Duplicate rows:
0

Data types:
Date                          datetime64[us]
City                                     str
State                                    str
Temperature_Max (°C)                 float64
Temperature_Min (°C)                 float64
Temperature_Avg (°C)                 float64
Humidity (%)                         float64
Rainfall (mm)                        float64
Wind_Speed (km/h)                    float64
AQI                                    int64
AQI_Category                             str
Pressure (hPa)                       float64
Cloud_Cover (%)                      float64
Year                                   int32
Month                                  int32
Day                                    int32
Day_of_Week                            int32
Day_Name                                 str
Temperature_Range (°C)               float64
Humidity_Temperature_Index           float64
dtype: object

Fi

,Date,City,State,Temperature_Max (°C),Temperature_Min (°C),Temperature_Avg (°C),Humidity (%),Rainfall (mm),Wind_Speed (km/h),AQI,AQI_Category,Pressure (hPa),Cloud_Cover (%),Year,Month,Day,Day_of_Week,Day_Name,Temperature_Range (°C),Humidity_Temperature_Index
0,2024-01-01,Mumbai,Maharashtra,32.5,18.0,25.2,77.6,0.0,3.3,259,Poor,1020.3,62.1,2024,1,1,0,Monday,14.5,1955.52
1,2024-01-01,Delhi,Delhi,25.4,10.7,18.1,84.1,0.0,9.0,130,Moderate,1008.4,46.0,2024,1,1,0,Monday,14.7,1522.21
2,2024-01-01,Bengaluru,Karnataka,37.2,30.8,34.0,49.0,3.7,6.6,54,Satisfactory,1008.0,61.3,2024,1,1,0,Monday,6.4,1666.00
3,2024-01-01,Chennai,Tamil Nadu,37.2,30.4,33.8,34.2,9.5,9.0,176,Moderate,993.4,70.0,2024,1,1,0,Monday,6.8,1155.96
4,2024-01-01,Kolkata,West Bengal,27.4,17.5,22.5,32.2,9.1,9.2,97,Satisfactory,1008.2,56.9,2024,1,1,0,Monday,9.9,724.50



Final columns:
['Date', 'City', 'State', 'Temperature_Max (°C)', 'Temperature_Min (°C)', 'Temperature_Avg (°C)', 'Humidity (%)', 'Rainfall (mm)', 'Wind_Speed (km/h)', 'AQI', 'AQI_Category', 'Pressure (hPa)', 'Cloud_Cover (%)', 'Year', 'Month', 'Day', 'Day_of_Week', 'Day_Name', 'Temperature_Range (°C)', 'Humidity_Temperature_Index']


In [14]:
# Save the cleaned and feature-engineered dataset

output_path = "../data/processed/climate_cleaned.csv"

df.to_csv(output_path, index=False)

print("Cleaned dataset saved successfully!")
print("File:", output_path)

Cleaned dataset saved successfully!
File: ../data/processed/climate_cleaned.csv


# Summary

## Data Cleaning

- The original climate dataset was loaded without modifying the original file.
- Dataset shape was verified as 7310 rows and 13 columns.
- No missing values were found.
- No duplicate rows were found.
- Date was converted from string format to datetime format.
- Logical validity checks were performed for temperature, humidity, rainfall, wind speed, AQI, pressure, and cloud cover.
- No invalid weather values requiring removal were found.
- Rainfall outliers detected using the IQR method were retained because extreme rainfall can represent genuine weather events.

## Feature Engineering

The following features were created from the Date column:

- Year
- Month
- Day
- Day_of_Week
- Day_Name

The following weather-based features were created:

- Temperature_Range (°C)
- Humidity_Temperature_Index

Highly redundant temperature-difference features were removed based on correlation analysis.

## Target Variable

AQI was retained as the target variable. AQI was not used to create any derived feature in order to avoid data leakage.

## Final Dataset

The final dataset contains 7310 rows and 20 columns.

The cleaned and feature-engineered dataset was saved as:

`data/processed/climate_cleaned.csv`

This dataset is ready to be used by Member 3 for model training.